controller/metrics/workqueue.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file is copied and adapted from k8s.io/kubernetes/pkg/util/workqueue/prometheus
+// which registers metrics to the default prometheus Registry. We require very
+// similar functionality, but must register metrics to a different Registry.
+
+// Metrics subsystem and all keys used by the workqueue.
+
+package metrics
```

Comment:
Instead of implementing own metrics provider we could use `Registry` from `sigs.k8s.io/controller-runtime/pkg/metrics` package:


pkg/metrics/registry.go
```go
import (
	controllermetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
)

type MetricsServer struct {
	// ...
	registry                controllermetrics.RegistererGatherer
	// ...
}

func NewAppRegistry(appLister applister.ApplicationLister, appFilter func(obj interface{}) bool, appLabels []string) controllermetrics.RegistererGatherer {
	registry := controllermetrics.Registry
	// ...
}


```

Reply:
I prefer to not reuse their metrics because it doesn't support customizing histogram buckets.
From what I can see in our environment the upper limit 10s is very small. We can see queue metrics and work metrics up to several minutes.

Reply:
Got it. It makes sense. I wish they make it possible to configure metrics properties but it is what it is. Added one more minor request about copyright.